substrate/primitives/runtime-interface/src/pass_by.rs

```diff
@@ -15,419 +15,667 @@
+#[cfg(substrate_runtime)]
+impl<T: codec::Encode> IntoFFIValue for PassByCodec<T> {
+	type Destructor = Vec<u8>;
 
-	/// Construct `Self` from the given `inner`.
-	fn from_inner(inner: Self::Inner) -> Self;
+	fn into_ffi_value(value: &mut Self::Inner) -> (Self::FFIType, Self::Destructor) {
+		let data = value.encode();
+		(pack_ptr_and_len(data.as_ptr() as u32, data.len() as u32), data)
+	}
 }
 
-/// The implementation of the pass by inner type strategy. The type that uses this strategy will be
-/// passed between wasm and the host by using the wrapped inner type. So, this strategy is only
-/// usable by newtype structs.
-///
-/// Use this type as associated type for [`PassBy`] to implement this strategy for a type. Besides
-/// that the `PassByInner` trait need to be implemented as well.
-///
-/// This type expects the type that wants to use this strategy as generic parameter `T` and the
-/// inner type as generic parameter `I`.
+/// Pass a `&[T]` into the host using the SCALE codec.
 ///
-/// [`PassByInner`](derive.PassByInner.html) is a derive macro to implement this strategy.
+/// This encodes a `&[T]` into a `Vec<u8>` using the SCALE codec and then
+/// passes a pointer to that byte blob and its length to the host,
+/// which then reads it and decodes back into `Vec<T>` and passes
+/// a reference to that (as `&[T]`) into the host function.
 ///
-/// # Example
-/// ```
-/// # use sp_runtime_interface::pass_by::{PassBy, Inner, PassByInner};
-/// struct Test([u8; 32]);
-///
-/// impl PassBy for Test {
-///     type PassBy = Inner<Self, [u8; 32]>;
-/// }
-///
-/// impl PassByInner for Test {
-///     type Inner = [u8; 32];
+/// Raw FFI type: `u64` (a fat pointer; upper 32 bits is the size, lower 32 bits is the pointer)
+pub struct PassSliceRefByCodec<T>(PhantomData<T>);
+
+impl<T> RIType for PassSliceRefByCodec<T> {
+	type FFIType = u64;
+	type Inner = T;
+}
+
+#[cfg(not(substrate_runtime))]
+impl<'a, T: codec::Decode> FromFFIValue<'a> for PassSliceRefByCodec<&'a [T]> {
+	type Owned = Vec<T>;
+
+	fn from_ffi_value(
+		context: &mut dyn FunctionContext,
+		arg: Self::FFIType,
+	) -> Result<Self::Owned> {
+		let (ptr, len) = unpack_ptr_and_len(arg);
+		let vec = context.read_memory(Pointer::new(ptr), len)?;
+		<Vec::<T> as codec::Decode>::decode(&mut &vec[..]).map_err(|error| format!(
+			"could not SCALE-decode '{}' when marshalling hostcall's arguments through the FFI boundary: {error}",
+			type_name::<Vec<T>>()
+		))
+	}
+
+	fn take_from_owned(owned: &'a mut Self::Owned) -> Self::Inner {
+		&*owned
+	}
+}
+
+#[cfg(substrate_runtime)]
+impl<'a, T: codec::Encode> IntoFFIValue for PassSliceRefByCodec<&'a [T]> {
+	type Destructor = Vec<u8>;
+
+	fn into_ffi_value(value: &mut Self::Inner) -> (Self::FFIType, Self::Destructor) {
+		let data = codec::Encode::encode(value);
+		(pack_ptr_and_len(data.as_ptr() as u32, data.len() as u32), data)
+	}
+}
+
+/// A trait signifying a primitive type.
+pub unsafe trait Primitive: Copy {}
```

Comment:
DQ: Why does this need to be unsafe?

Reply:
Hmm... good point, I think in one of the previous versions of the PR this made sense, but now I guess it doesn't. (:

I made it safe.